java/ql/src/semmle/code/java/dataflow/internal/ContainerFlow.qll

```diff
@@ -159,6 +159,44 @@ private predicate taintPreservingArgumentToQualifier(Method method, int arg) {
   method.(CollectionMethod).hasName("offer") and arg = 0
 }
 
+/**
+ * Holds if `method` is a library method that returns tainted data if its
+ * `arg`th argument is tainted.
+ */
+private predicate taintPreservingArgumentToMethod(Method method, int arg) {
+  method.getDeclaringType().hasQualifiedName("java.util", "Collections") and
+  (
+    method
+        .hasName(["checkedCollection", "checkedList", "checkedMap", "checkedNavigableMap",
+              "checkedNavigableSet", "checkedSet", "checkedSortedMap", "checkedSortedSet",
+              "enumeration", "list", "max", "min", "singleton", "singletonList", "singletonMap",
```

Comment:
I think we should exclude `singletonMap` here.  As long as we're representing collection flow with taint steps then we can't distinguish between keys and values in a map, so so far we've taken the approach that a tainted map means that the values are tainted, and that the keys can never be tainted.  This is of course a trade-of, but with this level of precision, I think it's reasonable to assume that map keys don't carry taint very often.

Reply:
OK